### PR TITLE
Add new example notebook for transfer learning

### PR DESCRIPTION
**Pull request by joelnkn**
Adding a notebook for transfer learning following the checklist in https://github.com/chemprop/chemprop/pull/772 (and https://github.com/chemprop/chemprop/issues/556 / https://github.com/chemprop/chemprop/pull/676).

### CODE REVIEW

**Comment by shihchengli** [comment]
Thanks for making this new example. It seems the notebook cannot be run successfully on MPS. Perhaps try running it on the CPU of your MacBook or on other devices to ensure that the output of each cell is normal.

File: examples/transfer_learning.ipynb [hunk 1]
```diff
@@ -0,0 +1,776 @@
...
+   "source": [
+    "results = trainer.test(mpnn, test_loader)\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Transfer learning with multicomponenent models\n",
+    "Multi-component MPNN models have individual MPNN blocks for each molecule it parses in one input. These MPNN modules can be indepently frozen for transfer learning."
```
Comment (KnathanM) [inline examples/transfer_learning.ipynb:635]:
```suggestion
    "Multi-component MPNN models have individual MPNN blocks for each molecule it parses in one input. These MPNN modules can be independently frozen for transfer learning."
```

**Comment by KnathanM** [comment]
The notebook looks good to me. Though reviewing this makes me think that we aren't handling transfer learning correctly with batch norm layers. The running averages continue to update during each forward pass of training even if the mean and scale gradients are turned off during the backwards pass. See the discussion [here](https://discuss.pytorch.org/t/how-to-freeze-bn-layers-while-training-the-rest-of-network-mean-and-var-wont-freeze/89736). Looks like Lightning has a way to handle this with a [callback](https://lightning.ai/docs/pytorch/stable/api/lightning.pytorch.callbacks.BaseFinetuning.html#lightning.pytorch.callbacks.BaseFinetuning.freeze) but I haven't looked into this yet.

**Comment by shihchengli** [comment]
I think whether to freeze the mean and variance in a batch normalization layer is task-dependent. If the new task is closely related to the original task, and especially if the data characteristics (such as distributions and feature scales) are similar, there might be less need to update the batch normalization parameters. Conversely, if the task or data is quite different, one might consider adapting these parameters. However, for cases where the parameters in MPNN layers are fixed, I would opt to freeze the batch normalization parameters to maintain consistency with the fixed features extracted by the MPNN.

**Comment by KnathanM** [comment]
I checked out your branch and verified that yes, setting `mpnn.bn.eval()` before passing the model to the trainer is enough to keep the batch normalization running stats from being updated. I thought that `trainer.fit` would run `model.train()` on the whole model, but somewhere lightning must save that the batch normalization is set to eval() and not change that. 

Can you also add the equivalent of `mpnn.bn.eval()` to our transfer learning logic in `cli/train.py` (somewhere around line 800)? This might also be needed in `cli/fingerprint.py`. There we use `with torch.no_grad()` but we may also need `model.bn.eval()`. Can you check if this is the case and make that change as well if needed?

**Review by JacksonBurns — DISMISSED** [review]
Following up on some offline conversations - it looks like the pre-trained model here has a `StandardScalar` on its target values, and we should therefore be re-scaling the fine-tuning data according to the same distribution seen in pre-training and stored in the `StandardScalar`.

**Comment by JacksonBurns** [comment]
> Following up on some offline conversations - it looks like the pre-trained model here has a `StandardScalar` on its target values, and we should therefore be re-scaling the fine-tuning data according to the same distribution seen in pre-training and stored in the `StandardScalar`.

Here is some demo code to do this, given some `pretrained_model` which has been loaded with `load_from_checkpoint` or something similar:
```python
pretraining_scaler = StandardScaler()
pretraining_scaler.mean_ = pretrained_model.predictor.output_transform.mean.numpy()
pretraining_scaler.scale_ = pretrained_model.predictor.output_transform.scale.numpy()
```

**Comment by KnathanM** [comment]
I'll also note that we need to fix this in the CLI

**Comment by KnathanM** [comment]
@JacksonBurns, does the notebook now include your suggested changes?

**Comment by JacksonBurns** [comment]
@KnathanM looks good, do we also need to show it for the MulticomponentMPNN

File: examples/transfer_learning.ipynb [hunk 2]
```diff
@@ -0,0 +1,836 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Transfer Learning / Pretraining\n",
+    "Transfer learning (or pretraining) leverages knowledge from a pre-trained model on a related task to enhance performance on a new task. In chemprop we can use pre-trained model checkpoints to initialize a new model and also freeze components of the new model as it is trained, as demonstrated in this notebook."
```
Comment (shihchengli) [inline examples/transfer_learning.ipynb:8]:
```suggestion
    "Transfer learning (or pretraining) leverages knowledge from a pre-trained model on a related task to enhance performance on a new task. In Chemprop, we can use pre-trained model checkpoints to initialize a new model and freeze components of the new model during training, as demonstrated in this notebook."
```

File: examples/transfer_learning.ipynb [hunk 3]
```diff
@@ -0,0 +1,836 @@
...
+    "mcmpnn"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 48,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "blocks_to_freeze = [0,1]  # a list of indices of the individual MPNN blocks to freeze before training."
```
Comment (shihchengli) [inline examples/transfer_learning.ipynb:785]:
```suggestion
    "blocks_to_freeze = [0, 1]  # a list of indices of the individual MPNN blocks to freeze before training."
```

File: examples/transfer_learning.ipynb [hunk 4]
```diff
@@ -0,0 +1,836 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": 46,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "chemprop_dir = Path.cwd().parent\n",
+    "checkpoint_path = chemprop_dir / \"tests\" / \"data\" / \"example_model_v2_regression_mol+mol.ckpt\"  # path to the checkpoint file. "
+   ]
+  },
```
Comment (shihchengli) [inline examples/transfer_learning.ipynb:714]:
```suggestion
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Change checkpoint model inputs here"
   ]
  },
```

**Review by shihchengli — APPROVED** [review]

**Review by JacksonBurns — CHANGES_REQUESTED** [review]
Should:
```python
for idx in range(frzn_ffn_layers):
    mpnn.predictor.ffn[idx].requires_grad_(False)
    mpnn.predictor.ffn[idx + 1].eval()
```
be:
```python
for idx in range(frzn_ffn_layers):
    mpnn.predictor.ffn[idx].requires_grad_(False)
    mpnn.predictor.ffn[idx].eval()  # <-- freeze and eval the same layer?
```

**Comment by shihchengli** [comment]
I think the eval is specifically for the dropout layer after the ffn layer. (see [here](https://github.com/chemprop/chemprop/pull/952#issuecomment-2206748102))

**Review by JacksonBurns — APPROVED** [review]
LGTM
